src/sql_generator.cpp

```diff
@@ -0,0 +1,342 @@
+#include <iostream>
+#include <logging.hpp>
+#include <string>
+
+#include "../includes/sql_generator.hpp"
+
+using duckdb::KeywordHelper;
+
+// Utility
+std::string compute_absolute_table_name(const table_def &table) {
+  std::ostringstream out;
+  out << KeywordHelper::WriteQuoted(table.db_name, '"') << "."
+      << KeywordHelper::WriteQuoted(table.schema_name, '"') << "."
+      << KeywordHelper::WriteQuoted(table.table_name, '"');
+  return out.str();
+}
+
+void write_joined(
+    std::ostringstream &sql, const std::vector<const column_def *> columns,
+    std::function<void(const std::string &, std::ostringstream &)> print_str) {
+  bool first = true;
+  for (const auto &col : columns) {
+    if (first) {
+      first = false;
+    } else {
+      sql << ", ";
+    }
+    print_str(col->name, sql);
+  }
+}
+
+// DuckDB querying
+bool schema_exists(duckdb::Connection &con, const std::string &db_name,
+                   const std::string &schema_name) {
+  std::ostringstream out;
+  out << "SELECT schema_name FROM information_schema.schemata WHERE "
+         "catalog_name="
+      << KeywordHelper::WriteQuoted(db_name, '\'')
+      << " AND schema_name=" << KeywordHelper::WriteQuoted(schema_name, '\'');
+
+  auto query = out.str();
+  mdlog::info("schema_exists: " + query);
+  auto result = con.Query(query);
```

Comment:
We should use a prepared statement + query parameters to avoid SQL injection here :-)

And same in most statement below. Probably only the `CREATE TABLE` & `CREATE SCHEMA` should just sanitize the schema name to make sure it is a valid identifier/name.

Reply:
I replaced concatenations with prepared statements in all SELECT statements, but the biggest concatenated queries are the ones that have to list individual columns; those cannot be created with prepared statement because the number of clauses needed will depend on the number of columns.
```
INSERT INTO "fivetran_test"."main"."books2384035451" 
SELECT * EXCLUDE (_fivetran_deleted, _fivetran_synced) 
FROM localmem.arrow_view 
ON CONFLICT (id ) DO UPDATE SET 
	"title" = excluded."title", 
	"magic_number" = excluded."magic_number"
```

Reply:
That should still work, no ? As you generate the query, add to a vector of an agnostic type and add in enumerate paramaters the size of the vector +1.
iirc in postgres everything actually becomes a string anyway in the params array so vector<string> should suffice.

Reply:
talked offline; this was about the older version where SELECT queries were getting concatenated.